### PR TITLE
stop function and code improvements

### PR DESCRIPTION
**Pull request by chrisnorman7**
Made the following modifications:
- Created a new class BaseEvent from which KeyboardEvent and MouseEvent are inherited.
- Added a stop method to the Hook class which gracefull stops it from listening.
- Removed the try-except blocks from the hook method of Hook so that exceptions in handlers can be caught the proper way.

### CODE REVIEW

**Comment by chrisnorman7** [comment]
Sorry for the terseness of the main comment... I've never made a PR before! :-)

Cheers for pyhooked, it's a really lovely bit of code, and very helpful to me personally!

File: pyhooked/__init__.py [hunk 1]
```diff
@@ -217,6 +221,10 @@ def __init__(self):
         self.mouse_is_hook = False
         self.keyboard_is_hook = True
 
+    def stop(self):
+        """Stopp this object from listening."""
```
Comment (emmatyping) [inline pyhooked/__init__.py]: There is a typo here. (Stop not Stopp)

**Comment by emmatyping** [comment]
Hello Chris, thanks so much for opening this PR! Glad you like PyHooked. First off, since this is your first PR (congrats on that), let me give you some pointers. If you are still making changes it is common to say somewhere "WIP" so that people know not to merge and it is a work in progress. Also, it is generally better to break up changes in multiple PRs, as it makes reviewing _much_ easier. (I still am happy to accept this PR if it looks good and works well).

Secondly, glad to see you are adding tests to your new stop function. That is always a good thing to do.

Thirdly, while it may be nice to allow the application handle an exception in the hook handler, the try finally is there to assure that the code calls the next windows hook. If it fails to do so this can cause all mouse and keyboard input to stop completely even after exiting Python (which is of course an issue).

Fourthly, I like the idea of base event, but not entirely sure of its purpose. If it is implemented, I think to avoid namespace pollution (there may be other `BaseEvent`s imported), it may be a good idea to name it something like `HardwareEvent`. 

Fifthly, could you describe what OEM_8 is a bit more? Why you added then removed it?

Sixth, (oh my so many points), the `warn_unrecognised` should default to `False`, as otherwise it is a backwards incompatible change (people who grab the latest and don't expect the output shouldn't get it).

**Comment by chrisnorman7** [comment]
Hi mate,

Yeah, sorry for the number of changes... I grabbed pyhooked, made the 
change I needed so it worked, then found myself refining it as the 
project I was using it in threw up things.

OEM_8 was the ` key on my keyboard, but then I realised that there were 
loads of keys which didn't work, and I didn't want to just start 
inventing keys, hence removing it and adding in the warn thing.

If exceptions cause keyboard and mouse input to stop (which I agree is a 
horrible thing), would it work to leave the finally in there and finally 
return normally?

Could easily wrap the self.handler(event) bit in a try block with no 
except... I think that's valid Python anyhow! :-)

As for tests, I was wondering about making a more robust set of tests, 
but I couldn't think of a way to do it. Anyway, test.py allows you to 
see what keys are called so you can use them in larger projects.

Cheers, and thanks for looking over the PR... I promise I'll stop making 
rapid-fire updates now! :-)

**Comment by chrisnorman7** [comment]
Wow... 8 commits... I have been a busy boy! :-)

**Comment by emmatyping** [comment]
Doing a try/finally would be optimal actually. It would allow catching exceptions and make sure we pass to the next hook.

RE tests, I was considering trying to do tests myself, but there isn't a really great way to do it. I suppose we could do a test to make sure that stop is actually stopping Hook, but eh, I am not sure it is worth the work. There are only so many tests to cover.

**Comment by chrisnorman7** [comment]
Actually... Now I made handler a method... that if statement is kind of 
redundant anyhow... I'll update the code and push again.

**Comment by chrisnorman7** [comment]
OK... I'm not exactly thrilled with that code... It's against my nature 
to let an error pass, but as you say there's no way around it, so this 
is the best I could come up with...

Only other thing I can think of is to stop the hook after an exception 
was caught, but that seems cherlish.

Let me know what you think.

Cheers,

**Comment by emmatyping** [comment]
Well, in my mind, you can still catch the exception, so while if you aren't doing a try/except in the user code it won't raise an error, if implemented correctly, the developer can still catch it and decide what to do if an exception is raised.

EDIT: so it looks like you are doing a try/except/finally. I think it would be better to do just try/finally so the above is true.

**Comment by chrisnorman7** [comment]
I guess... I just personally feel we shouldn't be bothering about the 
bad things that happen if the coder makes a mistake in their handler... 
But I definitely see why you do. Would be nice if there was some way to 
have exceptions registered in some kind of queue affair so the 1 in 100 
case where exceptions simply cannot be allowed to happen they could 
still be checked for.

Although it does occurr to me that if your handler is doing something 
nifty with ctypes for example, you could conceivably crash Python in 
which case your handler would screw up the user's system in any case.

Anyway, brain babble over. Hope you like the commit! :-)

**Comment by emmatyping** [comment]
Sorry, I changed my mind, I do think we can leave it to the user to catch their own exceptions. As long as the handler doesn't quit on exception, then everything should be fine. I have a different idea for how I want to handle some of the exceptions that could arise in `hook`, which I will add later. If you squash the previous commit, I'd be happy to merge.

**Pull request merged**